patroni/log.py

```diff
@@ -4,63 +4,144 @@
+                    new_handler = RotatingFileHandler(os.path.join(config['dir'], __name__))
+                handler = new_handler or self._log_handler
                 handler.maxBytes = int(config.get('file_size', 25000000))
                 handler.backupCount = int(config.get('file_num', 4))
             else:
-                if self.handler is None or isinstance(self.handler, RotatingFileHandler):
-                    add_handler = logging.StreamHandler()
-                handler = add_handler or self.handler
+                if self._log_handler is None or isinstance(self._log_handler, RotatingFileHandler):
+                    new_handler = logging.StreamHandler()
+                handler = new_handler or self._log_handler
 
-            oldlogformat = (self.config or {}).get('format', PatroniLogger.DEFAULT_FORMAT)
+            oldlogformat = (self._config or {}).get('format', PatroniLogger.DEFAULT_FORMAT)
             logformat = config.get('format', PatroniLogger.DEFAULT_FORMAT)
 
-            olddateformat = (self.config or {}).get('dateformat') or None
+            olddateformat = (self._config or {}).get('dateformat') or None
             dateformat = config.get('dateformat') or None  # Convert empty string to `None`
 
-            if oldlogformat != logformat or olddateformat != dateformat or add_handler:
+            if oldlogformat != logformat or olddateformat != dateformat or new_handler:
                 handler.setFormatter(logging.Formatter(logformat, dateformat))
 
-            if add_handler:
-                self.root_logger.addHandler(add_handler)
+            if new_handler:
+                with self._log_handler_lock:
+                    if self._log_handler:
+                        self._old_handlers.append(self._log_handler)
+                    self._log_handler = new_handler
 
-                if self.handler is not None:
-                    self.root_logger.removeHandler(self.handler)
-                    self.handler.close()
-
-                self.handler = add_handler
-
-            self.config = config.copy()
+            self._config = config.copy()
             self.update_loggers()
+
+    def _close_old_handlers(self):
+        while True:
+            with self._log_handler_lock:
+                if not self._old_handlers:
+                    break
+                handler = self._old_handlers.pop()
+            try:
+                handler.close()
+            except Exception:
+                _LOGGER.exception('Failed to close the old log handler %s', handler)
+
+    def run(self):
+        while True:
+            self._close_old_handlers()
+
+            record = self._queue_handler.queue.get(True)
+            if record is None:
+                break
+
+            self._log_handler.handle(record)
+            self._queue_handler.queue.task_done()
+
+    def shutdown(self):
+        self._queue_handler.queue.put_nowait(None)
```

Comment:
so the exception `queue.Full` is not important during shutdown or ?

Reply:
We can't handle it anyway.